### ISSUE TITLE
MethodHandle changes required for jdk17

### ISSUE THREAD

**Issue by keithc-ca** [eclipse-openj9/openj9#12524]
The openj9-staging branch of jdknext does not compile:
```
src/java.base/share/classes/java/lang/invoke/VarForm.java:37: error: cannot find symbol
import static java.lang.invoke.MethodHandleNatives.Constants.REF_invokeStatic;
                                                  ^
  symbol:   class Constants
  location: class MethodHandleNatives
src/java.base/share/classes/java/lang/invoke/VarForm.java:37: error: static import only from classes and interfaces
import static java.lang.invoke.MethodHandleNatives.Constants.REF_invokeStatic;
^
src/java.base/share/classes/java/lang/invoke/VarForm.java:74: error: cannot find symbol
        Class<?> erasedValue = MethodTypeForm.canonicalize(value, MethodTypeForm.ERASE);
                                                                                ^
  symbol:   variable ERASE
  location: class MethodTypeForm
src/java.base/share/classes/java/lang/invoke/VarForm.java:75: error: cannot find symbol
        Class<?>[] erasedCoordinates = MethodTypeForm.canonicalizeAll(coordinates, MethodTypeForm.ERASE);
                                                                                                 ^
  symbol:   variable ERASE
  location: class MethodTypeForm
src/java.base/share/classes/java/lang/invoke/VarForm.java:134: error: cannot find symbol
            .resolveOrNull(REF_invokeStatic, implClass, methodName, type);
                           ^
  symbol:   variable REF_invokeStatic
  location: class VarForm
```
I had hoped it was just a matter of defining some new stubs, but it seems that
* 8265135: Reduce work initializing VarForms

delays initialization of `VarForm.memberName_table` which leads to NPEs in the `VarHandle` constructor because (`setter` is null) here:
```
    this.fieldType = setter.parameterType(setter.parameterCount() - 1);
```
This will need help from someone more familiar with the interaction between openj9 and openjdk in this area.

**Comment by pshipton** [comment]
@babsingh @tajila fyi

**Comment by tajila** [comment]
We may need to switch over to ojdkmh earlier that anticipated, @babsingh will provide the details

**Comment by babsingh** [comment]
Updating OpenJ9 MHs to support the latest JDKnext changes is not feasible. We do not have the resources.

Proposed fix: https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/308

This will enable OJDK MHs for OpenJ9 JDKnext (JDK17) by default.
The JDK will compile successfully.
There will be functional failures in test targets, which can be disabled until they are fixed. Or we can allow failing tests if they can be excluded from the promotion criteria.
JIT optimization are still being implemented. So, the performance will be bad in the short term.
The functional and perf issues should be resolved by the cut-off date for the 0.27 release (7 Jun 21).

**Comment by pshipton** [comment]
Acceptance build running https://ci.eclipse.org/openj9/job/Pipeline-OpenJDK-Acceptance/507/

**Comment by pshipton** [comment]
I think the best approach is to exclude the failures until they are fixed. There is typically new jdknext material every night, resulting in an acceptance build. Somebody would have to look at each acceptance build and evaluate the failures to determine if the level could be manually promoted or if there were new unexpected failures. 

The other possibility is to leave the jdknext openj9 branch as it is now, at the last working state before enabling OMH. The openj9-staging branch now has it enabled. Once we consider openj9-staging to be "good enough", or when it passes the current test criteria (sanity.functional) it will be promoted to the openj9 branch.

**Comment by pshipton** [comment]
There are quite a few failures still. Also 300MB+ artifacts from each failed run.
```
 5 - - JCL_Test_2 	46 sec	
 7 - - JCL_Test_none_SCC_0 	45 sec	
 142 - - jsr335tests_0 	29 sec	
 143 - - jsr335tests_2 	7.8 sec	
 144 - - jsr335tests_4 	49 sec	
 146 - - jsr335tests_7 	1 min 7 sec	
 147 - - jsr335tests_none_SCC_1 	18 sec	
 148 - - jsr335tests_none_SCC_3 	4.4 sec	
 150 - - jsr335tests_none_SCC_7 	4.8 sec	
 214 - - SharedCPEntryInvokerTests_1 	8.9 sec	
 5 - - JCL_Test_0 	42 sec	
 6 - - JCL_Test_1 	41 sec	
 8 - - JCL_Test_none_SCC_1 	43 sec	
 59 - - jsr335tests_1 	58 sec	
 60 - - jsr335tests_3 	46 sec	
 61 - - jsr335tests_5 	22 sec	
 62 - - jsr335tests_none_SCC_0 	17 sec	
 63 - - jsr335tests_none_SCC_2 	4.2 sec	
 64 - - jsr335tests_none_SCC_4 	4.5 sec	
 65 - - jsr335tests_none_SCC_5 	4.6 sec	
 109 - - SharedCPEntryInvokerTests_0 	8.6 sec	
 110 - - SharedCPEntryInvokerTests_2 
```

**Comment by pshipton** [comment]
I've disabled the acceptance build for the moment, we can't run it regularly in this state.

**Comment by babsingh** [comment]
> Acceptance build running https://ci.eclipse.org/openj9/job/Pipeline-OpenJDK-Acceptance/507/

A number of test targets are failing due to the same issue. Tracker summary:

1. Tracked by https://github.com/eclipse/openj9/issues/11927:
    - [test_RevealDirect_TargetVarargsAdaptor](https://github.com/eclipse/openj9/blob/6a123933eec829a8ae2d955679dd03da54af13ed/test/functional/Java8andUp/src/org/openj9/test/java/lang/invoke/Test_MethodHandleInfo.java#L658)
    - **Note:** @fengxue-IS is investigating.

2. Tracked by https://github.com/eclipse/openj9/issues/11924:
    - [test335DefenderSupersendsAsmAsMethodHandles](https://github.com/eclipse/openj9/blob/6a123933eec829a8ae2d955679dd03da54af13ed/test/functional/Jsr335/src/org/openj9/test/jsr335/defendersupersends/asm/TestDefenderMethodLookupAsm.java#L109)
    - [testConflictingDefinitionsInSuperInterface](https://github.com/eclipse/openj9/blob/6a123933eec829a8ae2d955679dd03da54af13ed/test/functional/Jsr335/src/org/openj9/test/jsr335/defendersupersends/asm/TestDefenderMethodLookupAsm.java#L88)
    - **Note:** @fengxue-IS is investigating.

3. Tracked by https://github.com/eclipse/openj9/issues/11923:
    - `testJITServer_0`
    - **Note:** @dmitry-ten's https://github.com/eclipse/openj9/pull/12525 should fix the above failure.

4. Tracked by https://github.com/eclipse/openj9/issues/12544:
    - `jit_jitt_0`
    - `jit_jitt_1`
    - `jit_jitt_2`
    - `jit_jitt_3`
    - `JCL_Test_JITHelpers_0`
    - `JCL_Test_JITHelpers_1`
    - **Note:** This failure is new. @nbhuiyan is investigating.

5. Tracked by https://github.com/eclipse/openj9/issues/11528#issuecomment-811912451:
    - `SharedCPEntryInvokerTests_0`
    - `SharedCPEntryInvokerTests_1`
    - `SharedCPEntryInvokerTests_2`
    - **Note:** This failure happens intermittently, and can be disabled in the short term.

**Comment by pshipton** [comment]
I think what we should do is just run one of these builds per week for the time being, so we can monitor the list of failures, but not be overwhelmed by builds and failure artifacts. 

Atm since the latest changes were merged earlier today, the build is broken (unrelated to OMH https://github.com/eclipse/openj9/issues/12545), but Jason is looking and I don't expect it will take long to fix.

**Comment by JasonFengJ9** [comment]
https://github.com/eclipse/openj9/pull/12547 is created to address https://github.com/eclipse/openj9/issues/12545 though it requires `JDK17 adopts RI java.lang.String` (https://github.com/eclipse/openj9/pull/12209)

**Comment by babsingh** [comment]
Reopening, accidentally closed via  #12528.

**Comment by JasonFengJ9** [comment]
JDK17 `openj9-staging` branch has been enabled w/ OJDK MH, and it compiles successfully.

Closing..
